pysmt/formula.py

```diff
@@ -1099,17 +1095,22 @@ def normalize(self, formula):
             self._normalizer = FormulaContextualizer(self.env)
         return self._normalizer.walk(formula)
 
-    def _polymorph_args_to_tuple(self, args):
+    def _sorted_tuple(self, nodes):
+        return tuple(sorted(nodes, key=lambda p: p.node_id()))
+
+    def _polymorph_args_to_tuple(self, args, sort=False):
```

Comment:
In what cases do we want this NOT to be sorted? I'd expect this method is always called with `sort=True`, but I see a few instances that do not. Those instances are then not guaranteed to create the same formula.

Reply:
I believe in cases such as `StrConcat`, `BVConcat` we cannot sort the arguments.

In addition, at the moment I have not applied the sorting to all possible operators. e.g. I have omitted BV operators.
It should be possible to sort those as well. I can them.

I decided to add the extra optional argument to avoid duplicating the code to get the tuple from either a n-ary args or a single list argument.

An approach to avoid the extra arguments would be to define a function to get the list of arguments `_polymorph_args_to_lst(self, args)` and then define two variants `_args_to_tuple(self, args)` and `_args_to_sorted_tuple(self, args)`.

Reply:
> I believe in cases such as StrConcat, BVConcat we cannot sort the arguments.

You are right, makes sense, as we use `_polymorph_args_to_tuple` for all n-ary expressions.

> In addition, at the moment I have not applied the sorting to all possible operators. e.g. I have omitted BV operators. It should be possible to sort those as well. I can them.

It would be good to have all operators use the same approach. I am good if you want to do this as a follow-up though.

> I decided to add the extra optional argument

The extra argument is totally fine. As this is not semantic preserving in some cases (:point_up:), I also think the default is the correct one.